File: juno_rs/src/strategies/double_ma_2.rs
use super::{Signal, Strategy, StrategyMeta};
use crate::{
    genetics::Chromosome,
    indicators::{MAExt, MAParams, MA},
    Advice, Candle,
};
use juno_derive_rs::*;
use rand::prelude::*;
use serde::{Deserialize, Serialize};
use std::cmp::max;

#[derive(Chromosome, Clone, Debug, Deserialize, Serialize)]
#[repr(C)]
pub struct DoubleMA2Params {
    pub neg_threshold: f64,
    pub pos_threshold: f64,
    pub mas: (MAParams, MAParams),
}

fn mas(rng: &mut StdRng) -> (MAParams, MAParams) {
    loop {
        let (s, l) = (rng.gen_range(1..200), rng.gen_range(2..201));
        if s < l {
            return (rng.gen_ma_params(s), rng.gen_ma_params(l));
        }
    }
}
fn neg_threshold(rng: &mut StdRng) -> f64 {
    rng.gen_range(-1.0..-0.1)
}
fn pos_threshold(rng: &mut StdRng) -> f64 {
    rng.gen_range(0.1..1.0)
}

#[derive(Signal)]
pub struct DoubleMA2 {
    short_ma: Box<dyn MA>,
    long_ma: Box<dyn MA>,
    neg_threshold: f64,
    pos_threshold: f64,
    advice: Advice,
}

unsafe impl Send for DoubleMA2 {}
unsafe impl Sync for DoubleMA2 {}

impl Strategy for DoubleMA2 {
    type Params = DoubleMA2Params;

    fn new(params: &Self::Params, _meta: &StrategyMeta) -> Self {
        // Non-zero period is validated within indicator.
        let (short_ma, long_ma) = &params.mas;
        assert!(short_ma.period() < long_ma.period());
        assert!(params.pos_threshold > 0.0 && params.pos_threshold < 1.0);
        assert!(params.neg_threshold < 0.0 && params.neg_threshold > -1.0);

        Self {
            short_ma: short_ma.construct(),
            long_ma: long_ma.construct(),
            neg_threshold: params.neg_threshold,
            pos_threshold: params.pos_threshold,
            advice: Advice::None,
        }
    }

    fn maturity(&self) -> u32 {
        max(self.long_ma.maturity(), self.short_ma.maturity())
    }

    fn mature(&self) -> bool {
        self.long_ma.mature() && self.short_ma.mature()
    }

    fn update(&mut self, candle: &Candle) {
        self.short_ma.update(candle.close);
        self.long_ma.update(candle.close);

        if self.mature() {
            let diff = 100.0 * (self.short_ma.value() - self.long_ma.value())
                / ((self.short_ma.value() + self.long_ma.value()) / 2.0);

            if diff > self.pos_threshold {
                self.advice = Advice::Long;
            } else if diff < self.neg_threshold {
                self.advice = Advice::Short;
            }
        }
    }
}

File: juno/strategies/double_ma_2.py
import operator
from decimal import Decimal

from juno import Advice, Candle, CandleMeta, indicators
from juno.constraints import Int, Pair, Uniform
from juno.indicators import MA, Ema
from juno.inspect import get_module_type

from .strategy import Signal, Strategy, ma_choices


# Moving average moving average crossover.
class DoubleMA2(Signal):
    @staticmethod
    def meta() -> Strategy.Meta:
        return Strategy.Meta(
            constraints={
                ("short_period", "long_period"): Pair(Int(1, 100), operator.lt, Int(2, 101)),
                "neg_threshold": Uniform(Decimal("-1.000"), Decimal("-0.100")),
                "pos_threshold": Uniform(Decimal("+0.100"), Decimal("+1.000")),
                "short_ma": ma_choices,
                "long_ma": ma_choices,
            }
        )

    _short_ma: MA
    _long_ma: MA
    _neg_threshold: Decimal
    _pos_threshold: Decimal
    _advice: Advice = Advice.NONE

    def __init__(
        self,
        short_period: int,
        long_period: int,
        neg_threshold: Decimal,
        pos_threshold: Decimal,
        short_ma: str = Ema.__name__.lower(),
        long_ma: str = Ema.__name__.lower(),
    ) -> None:
        assert short_period > 0
        assert short_period < long_period

        self._short_ma = get_module_type(indicators, short_ma)(short_period)
        self._long_ma = get_module_type(indicators, long_ma)(long_period)
        self._neg_threshold = neg_threshold
        self._pos_threshold = pos_threshold

    @property
    def advice(self) -> Advice:
        return self._advice

    @property
    def maturity(self) -> int:
        return self._long_ma.maturity

    @property
    def mature(self) -> bool:
        return self._long_ma.mature

    def update(self, candle: Candle, _: CandleMeta) -> None:
        self._short_ma.update(candle.close)
        self._long_ma.update(candle.close)

        if self.mature:
            diff = (
                100
                * (self._short_ma.value - self._long_ma.value)
                / ((self._short_ma.value + self._long_ma.value) / 2)
            )

            if diff > self._pos_threshold:
                self._advice = Advice.LONG
            elif diff < self._neg_threshold:
                self._advice = Advice.SHORT
